integrate/adaptive.py

```diff
@@ -0,0 +1,119 @@
+import heapq
+
+
+def quad_vec(f, a, b, tol, quadrature, norm_func):
+    """
+    Adaptive quadrature of vector-valued function.
+
+    Parameters
+    ----------
+    f : callable
+        Vector-valued function f(x) to integrate
+    a : float
+        Initial point.
+    b : float
+        Final point.
+    tol : float
+        Absolute tolerance.
+    quadrature : callable
+        Underlying quadrature rule to use. Should be a function like
+        quadrature(f, a, b, norm_func) which returns an estimate of the integral and
+        and an estimate of the error.
+    norm_func: callable
+        Vector norm to use for error estimation.
+
+    Returns
+    -------
+    res : {float, array-like}
+        Estimate for the final result.
+    err : float
+        Error estimate for the result using the given norm.
+    """
+    global_est, global_err = quadrature(f, a, b, norm_func)
+
+    # Since tuples are compared based on their first element, making the first entry
+    # -err means that the top of the heap is always the interval with the largest error
+    intervals = [(-global_err, a, b, global_est)]
+
+    while global_err > tol:
+        interval = heapq.heappop(intervals)
+        neg_err_k, a_k, b_k, est_k = interval
+        err_k = -neg_err_k
+
+        m = (a_k + b_k) / 2
+
+        est_left, err_left = quadrature(f, a_k, m, norm_func)
+        est_right, err_right = quadrature(f, m, b_k, norm_func)
+
+        global_est = global_est - est_k + est_left + est_right
+        global_err = global_err - err_k + err_left + err_right
+
+        heapq.heappush(intervals, (-err_left, a_k, m, est_left))
+        heapq.heappush(intervals, (-err_right, m, b_k, est_right))
+
+    return global_est, global_err
+
+
+def ndquad_vec(f, ranges, tol, quadrature, norm_func):
+    """
+    Adaptive quadrature of an n-dimensional vector-valued function.
+
+    Wraps `quad_vec` to enable integration over multiple variables, it is not doing true
+    cubature.
```

Comment:
At this point in the docstring you probably want to give a proper math definition with displaymath ``.. math::` admonition.

Also need to define somewhere what is `n`, what is n-dimensional, what are the shapes of inputs and outputs.